### ISSUE TITLE
How do you use token balancing loss under pipeline parallelism

### ISSUE THREAD

**Issue by szhengac** [stanford-futuredata/megablocks#64]
Hi, I see there is[ balancing loss implementation](https://github.com/stanford-futuredata/megablocks/blob/main/megablocks/layers/moe.py#L30) with pipeline parallelism in moe.py. But I wonder how do you use it with Megatron-LM? It seem there is no example training code in the repository. With pipeline parallelism, we probably can only compute the backward of the balancing loss when the next stage passes the output gradient back.

**Comment by tgale96** [comment]
Hi! We have trained MoEs with pipeline parallelism in Megatron. You should be able to take one of our [training scripts](https://github.com/stanford-futuredata/megablocks/tree/main/exp/dmoe) and configure the pipeline parallelism arguments.

cc @deepakn94, who got pipeline parallelism to work with Megatron + MegaBlocks.

**Comment by tgale96** [comment]
They do not - you'll have to set the arguments to pass to Megatron, but it should work fine :)

**Comment by szhengac** [comment]
I see. It seems you are using your own fork for Megatron-LM. Can you please point out the lines of code that support routing balancing loss with pipeline parallelism? I think the official Megatron-LM main branch does not support that.

**Comment by tgale96** [comment]
You can see the changes we made in the top four commits from Deepak [here](https://github.com/NVIDIA/Megatron-LM/compare/main...stanford-futuredata:Megatron-LM:main). Hope this helps!
